sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala

```diff
@@ -154,8 +151,163 @@ object GpuScalar {
+                GpuColumnVector.typeConversionAllowed(elementView, elementType)
+              }
+            } else {
+              false
+            }
+          case _ => false // Unsupported type
+        }
+      case nonNested =>
+        GpuColumnVector.getNonNestedRapidsType(sType).equals(nonNested)
+    }
+  }
 }
 
+/**
+ * The wrapper of a Scala value and its corresponding cudf Scalar, along with its DataType.
+ *
+ * This class is introduced because many expressions require both the cudf Scalar and its
+ * corresponding Scala value to complete their computations. e.g. 'GpuStringSplit',
+ * 'GpuStringLocate', 'GpuDivide', 'GpuDateAddInterval', 'GpuTimeMath' ...
+ * So only either a cudf Scalar or a Scala value can not support such cases, unless copying data
+ * between the host and the device each time being asked for.
+ *
+ * This GpuScalar can be created from either a cudf Scalar or a Scala value. By initializing the
+ * cudf Scalar or the Scala value lazily and caching them after being created, it can reduce the
+ * unnecessary data copies.
+ *
+ * If a GpuScalar is created from a Scala value and is used only on the host side, there will be
+ * no data copy and no cudf Scalar created. And if it is used on the device side, only need to
+ * copy data to the device once to create a cudf Scalar.
+ *
+ * Similarly, if a GpuScalar is created from a cudf Scalar, no need to copy data to the host if
+ * it is used only on the device side (This is the ideal case we like, since all is on the GPU).
+ * And only need to copy the data to the host once if it is used on the host side.
+ *
+ * So a GpuScalar will have at most one data copy but support all the cases. No round-trip
+ * happens.
+ *
+ * Another reason why storing the Scala value in addition to the cudf Scalar is
+ * `GpuDateAddInterval` and 'GpuTimeMath' have different algorithms with the 3 members of
+ * a `CalendarInterval`, which can not be supported by a single cudf Scalar now.
+ *
+ * Do not create a GpuScalar from the constructor, instead call the factory APIs above.
+ */
+class GpuScalar private(
+    private var scalar: Scalar,
+    private var value: Option[Any],
+    val dataType: DataType, converted: Boolean) extends Arm with AutoCloseable {
+
+  assert(scalar != null || value.nonEmpty, "GpuScalar requires at least a value or a Scalar")
+  assert(value.isEmpty || !value.get.isInstanceOf[Scalar], "Value should not be Scalar")
+
+  /**
+   * Gets the internal cudf Scalar of this GpuScalar.
+   *
+   * This will not increase any reference count. So users need to close either the GpuScalar or
+   * the return cudf Scalar, not both.
+   */
+  def getBase: Scalar = {
+    if (scalar == null) {
+      scalar = GpuScalar.from(value.get, dataType, converted)
+    }
+    scalar
+  }
+
+  /**
+   * Gets the internal Scala value of this GpuScalar.
+   */
+  def getValue: Any = {
+    if (value.isEmpty) {
+      value = Some(GpuScalar.extract(scalar))
```

Comment:
It will return the value inside the option, not the option itself.

This GpuScalar can be created from either a `value` or a `Scalar`.   The undefined value means the GpuScalar is created from a cudf Scalar, and will be defined when calling the `getValue` for the first time.  The later calls to this `getValue` will not read the data from GPU again, since it has been already cached in the Some wrapper.

Use option for the `value` because `null` can not tell the difference between the null value and no value being specified at first.